ksqldb-engine/src/main/java/io/confluent/ksql/query/TransientQuerySinkProcessor.java

```diff
@@ -0,0 +1,170 @@
+          timestamp -> checkForQueryCompletion()
+      );
+    }
+  }
+
+  @Override
+  public void process(final Record<Object, GenericRow> record) {
+    final Optional<TopicPartition> topicPartition =
+        context
+            .recordMetadata()
+            .map(m -> new TopicPartition(m.topic(), m.partition()));
+
+    final boolean alreadyDone =
+        topicPartition.isPresent() && donePartitions.contains(topicPartition.get());
+
+    // We ignore null values in streams as invalid data and drop the record.
+    // If we're already done with the current partition, we just drop the record.
+    if (record.value() != null && !alreadyDone) {
+      queue.acceptRow(null, record.value());
+    }
+
+    topicPartition.ifPresent(this::checkForPartitionCompletion);
+  }
+
+  private void checkForPartitionCompletion(final TopicPartition topicPartition) {
+    if (endOffsets.isPresent()) {
+      final ImmutableMap<TopicPartition, Long> endOffsetsMap = endOffsets.get();
+      if (endOffsetsMap.containsKey(topicPartition)) {
+        final Map<TopicPartition, OffsetAndMetadata> currentPositions = getCurrentPositions();
+        checkCompletion(topicPartition, currentPositions, endOffsetsMap.get(topicPartition));
+      }
+    }
+  }
+
+  private void checkForQueryCompletion() {
+    if (endOffsets.isPresent()) {
+      final Map<TopicPartition, OffsetAndMetadata> currentPositions =
+          getCurrentPositions();
+
+      for (final Entry<TopicPartition, Long> end : endOffsets.get().entrySet()) {
+        checkCompletion(
+            end.getKey(),
+            currentPositions,
+            end.getValue()
+        );
+      }
+      // if we're completely done with this query, then call the completion handler.
+      if (ImmutableSet.copyOf(donePartitions).equals(endOffsets.get().keySet())) {
+        queue.complete();
+      }
+    }
+  }
+
+  private void checkCompletion(final TopicPartition topicPartition,
+      final Map<TopicPartition, OffsetAndMetadata> currentPositions,
+      final Long end) {
+
+    if (currentPositions.containsKey(topicPartition)) {
+      final OffsetAndMetadata current = currentPositions.get(topicPartition);
+      if (current != null && current.offset() >= end) {
+        donePartitions.add(topicPartition);
+      }
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private Map<TopicPartition, OffsetAndMetadata> getCurrentPositions() {
+    // It might make sense to actually propose adding the concept of getting the
+    // "current position" in the ProcessorContext, but for now we just expose it
+    // by reflection. Obviously, this code is pretty brittle.
+    try {
+      if (context.getClass().equals(ProcessorContextImpl.class)) {
+        final Field streamTask;
+        streamTask = ProcessorContextImpl.class.getDeclaredField("streamTask");
+        streamTask.setAccessible(true);
+        final StreamTask task = (StreamTask) streamTask.get(context);
+        final Method committableOffsetsAndMetadata =
```

Comment:
If the current position is referring to the source topic's position, could we get them via `ProcessorContext#recordMetadata()#offset()`? cc @AlanConfluent

Reply:
I tried that first, only to sadly be reminded that the `endOffsets` returned by the admin client are not the "offsets of the last records in the partitions", but rather "the offset _after_ the last record in the partition" aka "the offset that the _next produced_ record will get".

The reason I cracked the safe to get at this method here is because it specifically computes that position for the purpose of committing offsets. I.e., we don't commit the offsets of the records we have processed, we commit the offsets of the _next records_ we need to process. It might be tempting to just use `recordMetadata#offset() + 1` for this, but the reality is that it's not so trivial, thanks for skips in the offset sequence due to compaction, "invisible" offsets for transaction markers, and stuff like that.

The method I'm exposing here specifically accounts for all of those effects:
https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java#L435-L455

In addition to the above, we have another problem: upstream processors might filter out arbitrary records before they even reach this processor. Specifically, if you write a stream pull query that should only match one record, and there are more records after it in the topic, the query would never terminate, since the `process()` method would never see any offsets passed the one we're filtering for. That's why I also registered the punctuator, and the punctuator needs to be able to check "the current position" of the task as a whole, since it has no `recordMetadata` available to get the offset of.

Reply:
Thanks John, that makes sense!